pkg/provision/ec2.go

```diff
@@ -0,0 +1,203 @@
+	})
+	svc := ec2.New(sess)
+	return &EC2Provisioner{ec2Provisioner: svc}, err
+}
+
+// Provision deploys an exit node into AWS
+func (p *EC2Provisioner) Provision(host BasicHost) (*ProvisionedHost, error) {
+	image, err := p.lookupAMI(host.OS)
+	if err != nil {
+		return nil, err
+	}
+
+	groupID, name, err := p.securityGroup(host.Additional["inlets-port"])
+	if err != nil {
+		return nil, err
+	}
+
+	runResult, err := p.ec2Provisioner.RunInstances(&ec2.RunInstancesInput{
+		ImageId:      image,
+		InstanceType: aws.String(host.Plan),
+		MinCount:     aws.Int64(1),
+		MaxCount:     aws.Int64(1),
+		UserData:     &host.UserData,
+		NetworkInterfaces: []*ec2.InstanceNetworkInterfaceSpecification{
+			{
+				DeviceIndex:              aws.Int64(int64(0)),
+				AssociatePublicIpAddress: aws.Bool(true),
+				DeleteOnTermination:      aws.Bool(true),
+				Groups:                   []*string{groupID},
+			},
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	err = p.ec2Provisioner.WaitUntilInstanceRunning(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{runResult.Instances[0].InstanceId},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = p.ec2Provisioner.CreateTags(&ec2.CreateTagsInput{
+		Resources: []*string{runResult.Instances[0].InstanceId},
+		Tags: []*ec2.Tag{
+			{
+				Key:   aws.String("Name"),
+				Value: aws.String(*name),
+			},
+		},
+	})
+
+	return &ProvisionedHost{
+		ID:     *runResult.Instances[0].InstanceId,
+		Status: "",
+	}, nil
+}
+
+// Status returns the ID, Status and IP of the exit node
+func (p *EC2Provisioner) Status(id string) (*ProvisionedHost, error) {
+	var status string
+	s, err := p.ec2Provisioner.DescribeInstanceStatus(&ec2.DescribeInstanceStatusInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(s.InstanceStatuses) == 0 {
+		return nil, fmt.Errorf("cannot retrieve status for instance: %s", id)
+	}
+
+	if *s.InstanceStatuses[0].InstanceStatus.Status == "ok" {
+		status = ActiveStatus
+	}
+
+	d, err := p.ec2Provisioner.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	if err != nil {
+		return nil, err
+	}
+	if len(d.Reservations) == 0 {
+		return nil, fmt.Errorf("cannot describe instance: %s", id)
+	}
+
+	return &ProvisionedHost{
+		ID:     id,
+		Status: status,
+		IP:     aws.StringValue(d.Reservations[0].Instances[0].PublicIpAddress),
+	}, nil
+}
+
+// Delete removes the exit node
+func (p *EC2Provisioner) Delete(id string) error {
+	i, err := p.ec2Provisioner.DescribeInstances(&ec2.DescribeInstancesInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	if err != nil {
+		return err
+	}
+	groups := i.Reservations[0].Instances[0].SecurityGroups
+
+	_, err = p.ec2Provisioner.TerminateInstances(&ec2.TerminateInstancesInput{
+		InstanceIds: []*string{aws.String(id)},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = p.ec2Provisioner.WaitUntilInstanceTerminated(&ec2.DescribeInstancesInput{
```

Comment:
Prefer non-blocking for the event-loop

Reply:
OK, I will remove this.

Reply:
@alexellis, I forgot that this is actually required, the instance needs to be terminated before we can remove the security group.